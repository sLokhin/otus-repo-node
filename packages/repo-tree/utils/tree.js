const symbols = ['├', '│', '└'];
const space = '\u0020';
const newLine = '\u000A';

function buildTree(data) {
  const { name, items } = data;

  const root = String(name) ? `${name}${newLine}` : `Root${newLine}`;

  if (!hasChildItems(data)) return root;

  return root + buildBranch(items);
}

function buildBranch(items, prefix = '') {
  let branch = '';

  items.forEach((item, i) => {
    const isLastItem = i === items.length - 1;
    const nextSymbol = isLastItem ? symbols[2] : symbols[0];
    branch += prefix + nextSymbol + item.name + newLine;

    if (hasChildItems(item))
      branch += buildBranch(
        item.items,
        prefix + (isLastItem ? space : symbols[1])
      );
  });

  return branch;
}

function hasChildItems(item) {
  const { items } = item;
  return Array.isArray(items) && items.length > 0;
}

export { buildTree };
