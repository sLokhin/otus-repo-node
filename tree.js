const symbols = ['├', '│', '└'];
const space = '\u0020';
const newLine = '\u000A';

const data = {
  name: 1,
  items: [
    {
      name: 2,
      items: [{ name: 3 }, { name: 4 }],
    },
    {
      name: 5,
      items: [{ name: 6 }],
    },
  ],
};

function buildTree(data) {
  const { name, items } = data;

  const root = String(name) ? `${name}${newLine}` : `Root${newLine}`;

  if (!hasChildItems(data)) return root;

  return root + buildBranch(items);
}

function buildBranch(items, prefix = '') {
  let branch = '';

  items.forEach((item, i) => {
    const isLastItem = i === items.length - 1;
    const nextSymbol = isLastItem ? symbols[2] : symbols[0];
    branch += prefix + nextSymbol + item.name + newLine;

    if (hasChildItems(item))
      branch += buildBranch(
        item.items,
        prefix + (isLastItem ? space : symbols[1])
      );
  });

  return branch;
}

function hasChildItems(item) {
  const { items } = item;
  return Array.isArray(items) && items.length > 0;
}

console.log('Build tree function');
console.log(buildTree(data));
