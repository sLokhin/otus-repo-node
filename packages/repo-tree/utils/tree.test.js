import { buildTree, buildBranch, hasChildItems } from './tree.js';

const testData0 = {
  name: '',
  items: [],
};

const testData1 = {
  name: 'TestRoot1',
  items: [],
};

const testData2 = {
  name: 'TestRoot2',
};

const testData3 = {
  name: 'TestRoot3',
  items: [{ name: 'item1' }, { name: 'item2' }],
};

const testData4 = {
  name: 'TestRoot4',
  items: [
    {
      name: 2,
      items: [{ name: 3 }, { name: 4 }],
    },
    {
      name: 5,
      items: [
        {
          name: 6,
          items: [
            { name: 7 },
            { name: 8, items: [{ name: 10 }, { name: 11 }] },
            { name: 9 },
          ],
        },
      ],
    },
  ],
};

describe('Test snapshot', () => {
  it('simple tree with two branches', () => {
    const data = testData3;
    expect(buildTree(data)).toMatchSnapshot();
  });

  it('complex recursive tree', () => {
    const data = testData4;
    expect(buildTree(data)).toMatchSnapshot();
  });
});

describe('Test buildTree() function', () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('default root name', () => {
    const data = testData0;
    expect(buildTree(data)).toBe('Root\n');
  });

  it('zero children tree', () => {
    const data = testData1;
    expect(buildTree(data)).toBe('TestRoot1\n');
  });

  it('test data with no "items" property not throw error', () => {
    const data = testData2;
    expect(() => buildTree(data)).not.toThrow();
    expect(buildTree(data)).toBe('TestRoot2\n');
  });

  it('simple tree with two items', () => {
    const data = testData3;
    expect(buildTree(data)).toBe('TestRoot3\n├item1\n└item2\n');
  });
});

describe('Test buildBranch() function', () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('simple two items branch', () => {
    const data = testData3;
    expect(buildBranch(data.items)).toBe('├item1\n└item2\n');
  });

  it('non-array "items" throw error', () => {
    expect(() => buildBranch()).toThrow();
  });
});

describe('Test hasChildItems() function', () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('branch with child items returns "true"', () => {
    const branch = {
      name: 'branch0',
      items: [{ name: 3 }, { name: 4 }],
    };
    expect(hasChildItems(branch)).toBe(true);
  });

  it('branch with no child items returns "false"', () => {
    const branch = {
      name: 'branch0',
      items: [],
    };
    expect(hasChildItems(branch)).toBe(false);
  });

  it('branch with no "items" property not throw error', () => {
    const branch = {
      name: 'branch0',
    };
    expect(() => hasChildItems(branch)).not.toThrow();
    expect(hasChildItems(branch)).toBe(false);
  });

  it('branch with non-array "items" not throw error', () => {
    const branch = {
      name: 'branch0',
      items: 'item1 item2 item3',
    };
    expect(() => hasChildItems(branch)).not.toThrow();
    expect(hasChildItems(branch)).toBe(false);
  });
});
