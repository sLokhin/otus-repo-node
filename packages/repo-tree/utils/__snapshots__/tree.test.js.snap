// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test snapshot complex recursive tree 1`] = `
"TestRoot4
├2
│├3
│└4
└5
 └6
  ├7
  ├8
  │├10
  │└11
  └9
"
`;

exports[`Test snapshot simple tree with two branches 1`] = `
"TestRoot3
├item1
└item2
"
`;
